src/WebJobs.Script.WebHost/Management/InstanceManager.cs

```diff
@@ -89,16 +90,19 @@ public async Task<string> SpecializeMSISidecar(HostAssignmentContext context, bo
             return null;
         }
 
-        public bool StartAssignment(HostAssignmentContext context, bool isWarmup)
+        public bool StartAssignment(HostAssignmentContext context)
         {
             if (!_webHostEnvironment.InStandbyMode)
             {
                 _logger.LogError("Assign called while host is not in placeholder mode");
                 return false;
             }
 
-            if (isWarmup)
+            if (context.IsWarmupRequest)
             {
+                // Based on profiling download code jit-ing holds up cold start.
+                // Pre-jit to avoid paying the cost later.
+                Task.Run(async () => await Download(context.GetRunFromPkgContext()));
                 return true;
             }
             else if (_assignmentContext == null)
```

Comment:
The 26 msec JIT time in the profile comes from running the else part durign specialization. anything we can do here?
Or maybe we could manually add this to PGO file and see if it goes away from the profile?

System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1+AsyncStateMachineBox`1[System.Threading.Tasks.VoidTaskResult Microsoft.Azure.WebJobs.Script.WebHost.Management.InstanceManager+<>c__DisplayClass11_0+<<StartAssignment>b__1>d].MoveNext(class System.Threading.Thread)

Reply:
Thanks. I tried to invoke only the safe/sideeffect free api calls without touching any other components so it doesnt interfere with the actual cold start call later on.

the else part additionally includes
applying appsettings
switching from standby mode to specialized mode

Both of which can make the container dirty and so we might need a rollback for the fake calls sideeffects.

Yes, Lets try the PGO approach for this section to see if that helps and we could decide if we want to handle this in the warmup call or PGO.